Решение задачи

gwalka opened the pull request:
Тестами не покрывал, пока на кросс ревью

Commit:
Решение задачи
Тестами не покрывал, пока на кросс ревью

yom-yoom commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,32 @@
+
+@NoArgsConstructor
+@Data
+public class VacancyDto {
```
Было бы неплохо навесить на созданные тобой Dto (как классы так и рекорды) аннотацию @Builder из ломбока, так будет проще создавать эти объекты (в нашем случае упростит тестирование)

gwalka replied:
сделал

yom-yoom commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.List;
+
+@Controller
```
Мы используем @ RestController вместо обычного, ибо он возвращает Json объект

gwalka replied:
в самом деле)) сделал

yom-yoom commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,36 @@
+@RequiredArgsConstructor
+public class VacancyController {
+    private final VacancyService service;
+    private final VacancyMapper mapper;
```
Кажется, что здесь mapper не нужен

gwalka replied:
удалил спасибо, забыл

yom-yoom commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyCandidateDto.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public record VacancyCandidateDto(long count, String name, List<Candidate> candidates, TeamRole position) {
```
Думаю лучше возвращать лист id кандидатов , вместо самих кандидатов

gwalka replied:
не не, это DTO самой вакансии которая передается кандидатам. обычная VacancyDTO для проектов

Artem-doronin replied:
Считаю не совсем хорошо отдавать из контроллера сущность. Вот public VacancyCandidateDto updateVacancy(long vacancyId, @RequestBody VacancyDto vacancyDto) {
        return service.updateVacancy(vacancyId, vacancyDto);
    } считаю лучше отдать список id  чем список Entity

yom-yoom commented on src/main/java/faang/school/projectservice/filter/vacancy/VacancyFilter.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.stream.Stream;
+
+public interface VacancyFilter {
+    public boolean isApplicable(VacancyFilterDto filter);
```
public можно не писать

yom-yoom commented on src/main/java/faang/school/projectservice/mapper/VacancyCandidateMapper.java:
```diff
@@ -0,0 +1,8 @@
+import faang.school.projectservice.dto.vacancy.VacancyCandidateDto;
+import faang.school.projectservice.model.Vacancy;
+
+public abstract class VacancyCandidateMapper {
```
Кажется, тут пропущена аннотация

gwalka replied:
добавил спасибо

yom-yoom commented on src/main/java/faang/school/projectservice/filter/vacancy/VacancyNameFilter.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public Stream<Vacancy> apply(Stream<Vacancy> vacancies, VacancyFilterDto filter) {
+        if (!isApplicable(filter)) {
```
Возможно проще делать эту проверку в сервисе, ибо с каждым новым фильтром все больше кода будет дублироваться

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    private boolean existUserRole(Long userId) {
```
передается параметр userId, но не используется

gwalka replied:
он у меня передается в следующем приватном методе непосредственно при вызове самого метода, передавая туда userId я могу сверить его с добавляемым в список кандидатом

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+        return candidateMapper.toDto(vacancy);
+    }
+
+    private boolean validateCreatorRole() {
```
Возможно стоит сделать этот метод void, так как возвращаемая переменная не используется на его вызовах

gwalka replied:
точно, я хотел ее вызвать изначально в методе как if(!validateCreatorRole()) и кинуть исключение изначально, а сейчас посмотрел и исключение у меня там прописано, проще просто вызвать как метод

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+        vacancy.setStatus(VacancyStatus.OPEN);
+        updateMetaData(vacancy);
+        Vacancy savedVacancy = repository.save(vacancy);
+        project.getVacancies().add(savedVacancy);
```
А если лист вакансий проекта будет null?

iAzhdar replied:
а если и не будет null то то для чего вообще эта строчка если далее нет никакой работы с project

gwalka replied:
> А если лист вакансий проекта будет null?

что предлагаешь? сделать проверку и выбросить исключение?

gwalka replied:
> а если и не будет null то то для чего вообще эта строчка если далее нет никакой работы с project
у меня в условиях задачи указанно, что вакансия обязательно должна относиться к проекту, добавляя в вакансия проект, а вакансию добавляя в лист вакансий проекта я покрываю данное условие

yom-yoom replied:
Можно и так , а можно проверить есть ли вакансии и если нет, то просто инициализировать как new ArrayList() и уже в него добавить новую вакансию

gwalka replied:
угу, сделал спасибо

iAzhdar replied:
> > а если и не будет null то то для чего вообще эта строчка если далее нет никакой работы с project
> > у меня в условиях задачи указанно, что вакансия обязательно должна относиться к проекту, добавляя в вакансия проект, а вакансию добавляя в лист вакансий проекта я покрываю данное условие

В задаче вроде так "Вакансия обязательно относится к проекту", 
когда ты пишешь 
vacancy.setProject(project);
ты уже для вакансию связываешь с проектом
а тут - project.getVacancies().add(savedVacancy); ты в памяти просто сохраняешь лист для проекта не используя даже дальше это сохранение

gwalka replied:
> > > а если и не будет null то то для чего вообще эта строчка если далее нет никакой работы с project
> > > у меня в условиях задачи указанно, что вакансия обязательно должна относиться к проекту, добавляя в вакансия проект, а вакансию добавляя в лист вакансий проекта я покрываю данное условие
> 
> В задаче вроде так "Вакансия обязательно относится к проекту", когда ты пишешь vacancy.setProject(project); ты уже для вакансию связываешь с проектом а тут - project.getVacancies().add(savedVacancy); ты в памяти просто сохраняешь лист для проекта не используя даже дальше это сохранение

Вполне вероятно ты прав, но оставлю на суд Юры. Я согласен что сохраняю и не использую, но по сути в проекте же как должны эти листы заполняться с вакансиями и по логике при создании вакансии она должна улетать в лист проекту. По крайней мере человеку который будет делать проект сложней из вакуума заполнять лист и делать какой то метод что если проект создал вакансию засетить в лист и тп. В общем я в смятении

iAzhdar replied:
> > > > а если и не будет null то то для чего вообще эта строчка если далее нет никакой работы с project
> > > > у меня в условиях задачи указанно, что вакансия обязательно должна относиться к проекту, добавляя в вакансия проект, а вакансию добавляя в лист вакансий проекта я покрываю данное условие
> > 
> > 
> > В задаче вроде так "Вакансия обязательно относится к проекту", когда ты пишешь vacancy.setProject(project); ты уже для вакансию связываешь с проектом а тут - project.getVacancies().add(savedVacancy); ты в памяти просто сохраняешь лист для проекта не используя даже дальше это сохранение
> 
> Вполне вероятно ты прав, но оставлю на суд Юры. Я согласен что сохраняю и не использую, но по сути в проекте же как должны эти листы заполняться с вакансиями и по логике при создании вакансии она должна улетать в лист проекту. По крайней мере человеку который будет делать проект сложней из вакуума заполнять лист и делать какой то метод что если проект создал вакансию засетить в лист и тп. В общем я в смятении

когда ты сохраняешь вакансию ты указываешь ей projectId
в самой таблице вакансий есть внешний ключ на проект, а в проекте нету он джоином забирает там онтумэни
когда потом подтянешь проект он джоином должен забрать этот лист вакансий найдя у вакансии внешний ключ на себя
и в твоем случае вообще не обязательно проекту сетить этот лист если ты дальше с ним не работаешь)
ну я могу ошибаться, надеюсь Юра прокомментирует)

zaqqaz226 replied:
Что ж. Да, тут будет вытащено все через связки по id. Так что я тоже за то, что в данном случае не обязательно сетать этот лист.

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+        updateMetaData(vacancy);
+        Vacancy savedVacancy = repository.save(vacancy);
+        project.getVacancies().add(savedVacancy);
+        return candidateMapper.toDto(savedVacancy);
```
можно добавить логи на сообщение об успешном создании новой вакансии

gwalka replied:
сделал

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+        Stream<Vacancy> vacancyStream = vacancyList.stream();
+        return filterVacancies(vacancyStream, filter)
+                .map(candidateMapper::toDto)
+                .collect(Collectors.toList());
```
Тут можно просто заменить на toList()

yom-yoom commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    private void validateCandidateAdding(Vacancy vacancy, Candidate candidate) {
+        if (existUserRole(new Candidate().getUserId())) {
```
Можно ли гарантировать, что у только что созданного кандидата без параметров будет id, тут не совсем понял

gwalka replied:
кандидат по сути является юзером, просто называется как кандидат в контексте вакансии. Добавляемый кандидат является юзером, мы получаем его Id и и сверяем в методе existUserRole возможно ли его добавить в данный список

Commit:
VacancyControllerTest +
покрытие замечаний из кросс ревью

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,33 @@
+    private String name;
+    private String description;
+    private TeamRole position;
+    private Project project;
```
мне кажется тебе тут стоит поменять Project project на Long projectId.
Дальше комменты  в service)

gwalka replied:
done

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+    public VacancyCandidateDto createVacancy(VacancyDto vacancyDto) {
+        validateCreatorRole();
+        Vacancy vacancy  = mapper.toEntity(vacancyDto);
+        Long projectId = vacancyDto.getProject().getId();
```
здесь ты тогда можешь не дергать целый проджект, а обратится только к id из vacancyDto

WhatIneedtoDO replied:
да и в принципе у тебя надобность в этой переменной отпадет

gwalka replied:
в целом отпадает в таком случае, но возникает вопрос. Мы ведь пытаемся сопоставить между собой принадлежность вакансии к проекту, чтобы между собой их сопоставить и правильно записать. Тоесть мы берем прям из вакансии айди проекта и это вроде как гарантирует нам что мы далее сетим все верно. Ну или я просто запутался

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+        validateCreatorRole();
+        Vacancy vacancy  = mapper.toEntity(vacancyDto);
+        Long projectId = vacancyDto.getProject().getId();
+        Project project = projectRepository.findById(projectId)
```
это можно сделать как if(в плане того, что к существующему ли проекту создана вакансия), но с другой стороны и exception должен все остановить, так что это 50\50

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+        Project project = projectRepository.findById(projectId)
+                .orElseThrow(() -> new DataValidationException("Проект не найден"));
+        mapper.updateVacancyFromDto(vacancyDto, vacancy);
+        vacancy.setProject(project); //todo;валидация?
```
можно будет все засетать в одну строку

gwalka replied:
Сетятся ведь поля по умолчанию кто что захотел тот то и поставил. Вот я решил что после того как все засечено маппером, автоматически вакансия приобретает статус Open, ибо там есть другие состояния, автоматически сетится проект который принадлежит вакансии. Затем валидируется информация указанная по условно по желанию. А после обновляется информация во сколько обновлена вакансия и кем. Либо я снова запутался)

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+        updateMetaData(vacancy);
+        Vacancy savedVacancy = repository.save(vacancy);
+        log.info("Вакансия {} успешно создана.", savedVacancy.getName());
+        if (project.getVacancies() == null) {
```
Думаю стоит добавить сюда еще проверку .contains, зачем тебе добавлять вакансию, если она уже есть у этого проекта?

gwalka replied:
а ведь айдишник вакансии генирируется автоматически и не может одна и та же вакансия дважды попасть в лист, нет? мы добавляем ее только при создании

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+        Vacancy targetVacancy = repository.findById(vacancyId)
+                .orElseThrow(() -> new DataValidationException("Вакансия не найдена"));
+
+        mapper.updateVacancyFromDto(vacancyDto, targetVacancy);
```
Крутое решение использовать обновление вакансии через маппер)

gwalka replied:
> Крутое решение использовать обновление вакансии через маппер)

как мило что ты свой совет назвал моим решением =D

WhatIneedtoDO commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,181 @@
+        return true;
+    }
+
+    private boolean hiredCandidates(Vacancy vacancy) {
```
может в валидации стоит использовать DTO ? тебе же с контроллера прилетает дто, которое ты валидируешь, а после уже маппишь в сущность

gwalka replied:
переделал, вызов и связанные с ним методы

Commit:
FilterNameTest+
FilterPositionTest+
правки в VacancyDto и FilterDto
исправление пустого листа в UserService

Commit:
Методы полностью покрыты тестами
В результате тестов изменены методы validateAndAddNewCandidates
addAndValidateCandidate
validateCreatorRole

zaqqaz226 commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,34 @@
+public class VacancyController {
+    private final VacancyService service;
+
+    public VacancyCandidateDto createVacancy(@RequestBody VacancyDto vacancyDto) {
```
Давай добавим уже маппинги на вызовы тоже

zaqqaz226 commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
```
Пора добавлять маппинги )

zaqqaz226 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,183 @@
+        TeamMember creator = memberRepository.findById(
+                userContext.getUserId()).orElseThrow(() ->
+                new DataValidationException("Пользователь не найден."));
+        List<TeamRole> creatorRoles = creator.getRoles();
```
Следующие 6 строк можно упростить перенеся стрим creator.getRoles().stream() сразу в if и сделав в этом стриме нужные проверки через ||

zaqqaz226 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,183 @@
+        }
+
+        Vacancy targetVacancy = repository.findById(vacancyId)
+                .orElseThrow(() -> new DataValidationException("Вакансия не найдена"));
```
Дополни ошибку, чтобы мы могли идентифицировать проблему. Например "Вакансия с id: {} не найдена"

zaqqaz226 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,183 @@
+        Vacancy vacancy  = mapper.toEntity(vacancyDto);
+        Long projectId = vacancyDto.getProjectId();
+        Project project = projectRepository.findById(projectId)
+                .orElseThrow(() -> new DataValidationException("Проект не найден."));
```
Дополни ошибку, чтобы мы могли идентифицировать проблему. Например "Проект с id: {} не найден"

zaqqaz226 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,183 @@
+        return true;
+    }
+
+    private void validateVacancyClose(VacancyDto vacancyDto) {
```
validateVacancyClose -> validateVacancyIsClosed или как-то так. Чтобы название было более логичное

gwalka replied:
точно? этот метод валидирует возможно ли закрыть вакансию. validateVacancyCanBeClosed?

zaqqaz226 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,183 @@
+                vacancyFilters.stream()
+                        .filter(filter -> filter.isApplicable(filters))
+                        .toList();
+        if (applicableFilters.isEmpty()) {
```
У тебя тут по логике либо фильры пустые, либо они не пустые и мы пойдем их проверять, так что лучше сделать после этого ифа else, тем самым у тебя будет всего 2 выхода из метода. Если у тебя логика либо - либо, то для наглядности, читаемости и расширяемости делаем через if-else

gwalka replied:
я вообще убрал if else, мне при таком подходе идея подсказала, что можно избавиться от ветвлений.

zaqqaz226 commented on src/test/java/faang/school/projectservice/service/VacancyServiceTest.java:
```diff
@@ -0,0 +1,380 @@
+        assertEquals(1L, result.id());
+        assertEquals("Designer", result.name());
+
+        verify(projectRepository, times(1)).findById(1L);
```
Давай сначала verify, а потом assert во всех тестах. Чисто по логике, мы сначала проверяем все ли правильно дергается, а потом уже сравниваем результаты.

zaqqaz226 commented on src/test/java/faang/school/projectservice/vacancyfiltertest/VacancyNameFilterTest.java:
```diff
@@ -0,0 +1,96 @@
+    @BeforeEach
+    public void setUp() {
+        filter = new VacancyNameFilter();
+        filterDto = VacancyFilterDto.builder().namePattern("Java").build();
```
Красоты ради, перенеси строки чтобы было 
VacancyFilterDto.builder()
.namePattern("Java")
.build();

Commit:
Controller добавлены маппинги
Service из фильтра убран is-else, изменен лог выдачи некоторых ошибок6

zaqqaz226 commented:
Ответил про тот лист. Поправь и давай закатывать задачку.

Commit:
убрал сэт вакансии из листа проекта
